README.md
# DEATH AND DECAY
## Introduction
- Welcome to Death and Decay, a monster/weapon replacer mod. If you liked Hard Doom and mods of that type, then you might like this mod. If not, it's still alright, this mod is a bit different compared to the others.
- The very first difference is that your character will constantly be evolving and improving in terms of both the stats and weapon efficiency. You'll be earning credits through monsters kills or completing a few special tasks. (These will be mentioned below) You can later use the hard earned cash to buy some very valuable items.
- This mod incorporates the popular survival features used in servers nowadays such as the revenant tracer fix, shared keys etc.
- Make sure you bind a key to the menu of DnD. It has lots of information regarding your weapons, stats, loadout, the shop etc. A must have!

## Screenshots
<img src="http://i.imgur.com/7Z9BpOY.png"/>
<img src="http://i.imgur.com/dOPqpvL.png"/>

## Releases
Stable builds are located at [Releases section](https://github.com/Doruk-Aksoy/DnD/releases).
Instructions on git building are located [here](./BUILDING.md).

## Support
For any questions concerning the mod, join our [discord](https://discord.gg/Gffvznu).
### Known Bugs
- DM only weapons/items/whatever will still spawn in Coop/Survival. Zandronum bug, don't bother me about this.
### Zandronum account
*  Required for Hardcore/Softcore modes.
1. Make an account here: [Click](http://auth.zandronum.com/)
2. When you are in game for a hardcore server, do "login username password". (The same name and password you entered for 1)
3. When the game prompts you to load your character, you can do "loadme x" where x is your character id.
Remember, your character is NOT saved until the map is officially over. This means, if you spectate at any point before the map ends, your progress is gone. You die, your character is entirely wiped out.
### Creating Server
- Add the pk7 in the following order (on Linux, use proper casing - if building from source using build.bat/sh, remember to use the versioned files):
```
map packs...
dnd.pk7
dnd_monsters.pk7
dnd_onlyammo.pk7
```

- Flags for servers:

```
DMFlags = 3227652
DMFlags2 = 256
DMFlags3 = 1104
CompatFlags2 = 2
```

- **BIG WARNING:** Remove death exit only maps from rotations in hardcore mode! Your characters can die and will be lost!

DnD/CHANGELOG.txt
==== DND 2022 APRIL 1ST EDITION v2.5 PATCH NOTES ====

* Added NFTs. (See Shop)
* Added new perk "Brutality".
	- 5% increased melee damage and 8% increased melee range per rank.
	- "Sharpshooting" no longer affects melee weaponry.
* You gain a mastery when you allocate maximum points (10) to a perk!
	- Sharpshooting: Dexterity gives 4% damage per point instead of 3%.
	- Brutality: Melee damage and range increase also applies to projectiles fired from melee weapons.
	- Endurance: Damage going to armor has 10% chance to be completely negated.
	- Wisdom: Gain experience from kills regardless of the killer or the range limit.
	- Greed: Gain credit from kills regardless of the killer or the range limit.
	- Medic: Convert 10% of healing pickups to extra armor.
	- Munitionist: Enemies have 2% chance to drop ammo tokens, giving 10% ammo to a random weapon you own.
	- Deadliness: Every 7th hit you land within four seconds is a guaranteed critical strike.
	- Savagery: 100% to crit multiplier if you haven't dealt a critical strike in 5 seconds.
	- Luck: When an item drops there's a 25% chance to get an additional item drop of its type.
* You can now invest into a research to increase its odds of dropping.
	- All research icons, cost etc. are visible from the start, but you still can't research them without the blueprints.
	- Until you get the blueprint, you can invest to add 1% chance to drop for a research per investment.
	- Investment uses credits.
	- You invest a base amount of 1000, and an additional 5% + 15% of the budget cost of that research per investment on top of your previous investment cost.
		+ Formula basically becomes: 
		+ Example: Research cost is 50k. You invest once, its 1000. The second time its 1125. Third time its 1406. ((1125 + (100 + (5 + 15*50/100) * 2)) / 100)
		+ Please note that precision errors apply due to integer math, so expected values may vary to be lower than shown in this example.
	- Maximum is (obviously) 100% drop chance.
	- If you pick the blueprint, you can no longer invest as it's pointless.
	- The drop chance is calculated on YOUR investment, other players' investments do not matter. This includes the hit that kills a monster or chests opened by players.
	- You can't invest in research that is obtained from specific monsters, through their unique drops or otherwise.
* Chests can now drop from the following items with the following chance:
	- Stimpacks: 0.25%
	- Medkit: 0.5%
	- Armors: 1%
	- SoulSphere, Megasphere: 10%
* Chests have a maximum amount that can spawn in map now scaling with difficulty:
	- Very Easy: 10
	- Easy: 13
	- Medium: 18
	- Hard: 24
	- Very Hard: 35
* You can pick up to 25 backpacks now down from 55 to increase your ammo capacities.
* "Demon Heart" nerfed.
	- Ammo cap of 2 instead of 5.
	- Damage reduced by 20%.
	- Distance halved.
	- Angle of scan reduced to 54 degrees from 90.
	- Drop chance halved to 5% from 10%.
* "Book of the Dead" range doubled.
* Charms now have level requirements. You can't equip the charm unless your level matches the level shown on the top left corner of the box.
* You can no longer use multiples of the same unique charms. Due to this, some charms a changed.
* "Well of Power" charm now grants:
	- Amplifies modifier values of small charms by 1.25 - 2.
	- Increases damage taken by 15-30%.
	- Can use maximum of 1-3 small charms.
* "Ancient Gemstone" buffed:
	- 5 - 20 all stats up from 5 - 8.
	- Chance to heal 3 - 8 from 3 - 5.
* "Deathspark" buffed:
	- Increased lightning damage to 5 - 60 from 5 - 45.
* "Oakheart" buffed:
	- Damage per 65-40 health instead of 75-50 health.
* "Pelletstorm" buffed:
	- 50-125% more pellets from 50-100%.
	- 180-300% slot3 damage from 180-250%.
* "Lifeleech" buffed:
	- 6-17.5% movespeed from 6-12%.
	- 3-12 flat damage from 3-10.
* "Shellshock" buffed:
	- Now gives a new mod 1-3 damage to shotguns per shotgun owned.
* "Rebounder" base damage increased to 30 from 25.
* "Rebounder" primary fire gives 3 overheat instead of 4.
* "Rebounder" altfire now uses 18 overheat down from 24.
* "Rebounder" primary fire overheat thresholds lowered to 15 - 30 - 45 - 60 from 24 - 48 - 72 - 96.
* "Rebounder" overheated shots have angle spread of 10 degrees starting from the 2nd threshold.
* "Jade Wand" primary attack damage per pellet reduced to 15 from 20.
* "Jade Wand" altfire cost increased to 8 from 5, and cooldown increased to 8 seconds from 5 seconds.
* "Jade Wand" corpse explosion damage reduced to 50 from 100.
* "Acid Rifle" magsize increased to 32 from 28.
* "Acid Rifle" poison alt spreads more and inflicts poison.
* "Acid Rifle" poison bolt speed increased to 56 from 40.
* "Blade Vortex" no longer grants damage reduction.
* Grenade Launcher and Mercury Launcher have swapped places. Now you need a research to obtain Mercury Launcher.
* Price of Grenade Launcher is now 2500 credits. Mercury Launcher is at 6800. Research cost is 65k.
* Charm modifier ranges can be toggled to be shown now from Menu "Help" section. You can also bind a key to do this without having to go to the menu, under "DnD Keys" section.
* "Luck" perk now affects rolls on regular charms, with 5% chance per point.
* "Strength" no longer gives health and armor cap bonuses.
* Culling effects buffed to 15% from 10%.
* Added a custom scoreboard that acts as a replacement for the end of level screen.
* Removed "X players alive" and "Last Player Alive" text.
* Hobo mugshot graphics updated.
* "Bloodfiend Spine" altfire can heal you now for each pellet hit for 1% of your health cap.
* Improved wording on Punisher level 5 perk. Buffed level 50 perk.
	- Level 5 perk now says "Being on a killing spree" instead of "Being on at least cruel killing spree".
	- Level 50 perk now gives 4% damage increase on that map per 10 kills instead of 50. (Caps at 400% still, so max 1000 kills required for max damage instead of 5000)
* Killing sprees start triggering for every 10 monster killed instead of 15.
* Your budget is now shown in the stat page.

acs/DnD_Ammo.h
#ifndef DND_AMMO_IN
#define DND_AMMO_IN

#include "DnD_Stat.h"

#define DND_MAX_BACKPACK 25
#define DND_BACKPACK_RATIO 20

enum {
	DND_AMMOSLOT_CLIP,
	DND_AMMOSLOT_SHELL,
	DND_AMMOSLOT_ROCKET,
	DND_AMMOSLOT_CELL,
	DND_AMMOSLOT_SOULS,
	DND_AMMOSLOT_MAGAZINE,
	DND_AMMOSLOT_SPECIAL,
	DND_AMMOSLOT_TEMPORARY
};

#define MAX_SLOTS (DND_AMMOSLOT_SOULS + 1)
#define MAX_AMMOTYPES_PER_SLOT 11

enum {
	DND_AMMO_SECONDARYISCLIP = 1
};

enum {
	// category 2
	AMMO_CLIP = 0,
	AMMO_RUBY,
	AMMO_VIPER,
	AMMO_DEMONSEAL,
	AMMO_RIOT,
	AMMO_ACID,
	AMMO_FUSIONCELL,
	AMMO_INCINERATOR,
	AMMO_DESOLATOR,
	AMMO_EBONY1,
	AMMO_EBONY2,
	
	// category 3
	AMMO_SHELL = 0,
	AMMO_EMERALDMANA,
	AMMO_HELLSMAWAMMO,
	AMMO_PCANNON,
	AMMO_NITROGENCANISTER,
	AMMO_DEMONBLOOD,
	AMMO_CHARON,
	AMMO_HADES,
	AMMO_FLAYER,
	AMMO_EXSHELL,
	AMMO_SLAYER,
	
	// category 4
	AMMO_ROCKET = 0,
	AMMO_FLAKSHELL,
	AMMO_METEOR,
	AMMO_HEAVYGRENADE,
	AMMO_GRAVDIS,
	AMMO_GRENADE,
	AMMO_HMISSILE,
	AMMO_SEDRIN,
	
	// category 5
	AMMO_CELL = 0,
	AMMO_DEVASTATOR,
	AMMO_EVERICE,
	AMMO_FUEL,
	AMMO_LIGHTNING,
	AMMO_NAIL,
	AMMO_LAVA,
	AMMO_ION,
	AMMO_THUNDER,
	AMMO_GAUSS,
	
	// category 6 (soul weapons -- only 1 ammo type here)
	AMMO_SOUL = 0
};

#define AMMOSLOT_CLIP_BEGIN AMMO_CLIP
#define AMMOSLOT_CLIP_END AMMO_EBONY1
#define AMMOSLOT_CLIP2_BEGIN AMMO_EBONY2
#define AMMOSLOT_CLIP2_END AMMO_EBONY2
#define AMMOSLOT_SHELL_BEGIN AMMO_SHELL
#define AMMOSLOT_SHELL_END AMMO_EXSHELL
#define AMMOSLOT_SHELL2_BEGIN AMMO_SLAYER
#define AMMOSLOT_SHELL2_END AMMO_SLAYER
#define AMMOSLOT_ROCKET_BEGIN AMMO_ROCKET
#define AMMOSLOT_ROCKET_END AMMO_SEDRIN
#define AMMOSLOT_CELL_BEGIN AMMO_CELL
#define AMMOSLOT_CELL_END AMMO_GAUSS

// don't make tables with strings and ints... weird fuckups happen
typedef struct {
	int initial_capacity;
	int container_value;
} ammo_info_T;

#define AMMOINFO_ICON 0
#define AMMOINFO_NAME 1
str AmmoInfo_Str[MAX_SLOTS][MAX_AMMOTYPES_PER_SLOT][2] = {
	// category 1 -- clip box
	{
		{		"CLIPA0",			"Clip"					},
		{		"RBAMA0",			"RubyAmmo"				},
		{		"C_H2A1",			"ViperAmmo"				},			
		{		"DEMSICO",			"DSealAmmo"				},			
		{		"RIOTJ0",			"RiotgunShell"			},
		{		"D97AA0",			"AcidAmmo"				},
		{		"FBATA0",			"FusionCell"			},
		{		"INCRAMMO",			"IncineratorAmmo"		},
		{		"DES4Y0",			"DesolatorAmmo"			},
		{		"EBOPA0",			"EbonyAmmo"				},
		{		"EBOPE0",			"EbonyAmmoX"			}
	},
	// category 2 -- shell box
	{
		{		"SHELA0",			"Shell"					},
		{		"EMEAA0",			"EmeraldMana"			},
		{		"HMAWA1",			"HellsMawAmmo"			},
		{		"PCNIC1",			"PCanAmmo"				},
		{		"D97A1",			"NitrogenCanister"		},
		{		"DBLUD1",			"DemonBlood"			},
		{		"CHARAMMO",			"CharonAmmo"			},
		{		"STRGY0",			"HadesAmmo"				},
		{		"CBWAA0",			"FlayerAmmo"			},
		{		"ESHAA0",			"ExplodingShell"		},
		{		"SLAYAM01",			"SlayerAmmo"			}
	},
	// category 3 -- rocket box
	{
		{		"ROCKA0",			"RocketAmmo"			},
		{		"FSHLA0",			"FlakShell"				},
		{		"LAAM1",			"MeteorAmmo"			},
		{		"HEGRA0",			"HeavyGrenades"			},
		{		"GRAVD0",			"GravdisAmmo"			},
		{		"GBUNA0",			"Grenades"				},
		{		"AHRLA0",			"MISAmmo"				},
		{ 		"SEDCRY",			"SedrinAmmo"			},
		{		"",					""						}
	},
	// category 4 -- cell box
	{
		{		"CELLA0",			"Cell"					},
		{		"DEVIAM0",			"DevastatorAmmo"		},
		{		"EICEAM1",			"EverIce"				},
		{		"FUAMA0",			"Fuel"					},
		{		"D98AB1",			"LightningCell"			},
		{		"NPAKA0",			"NailgunAmmo"			},
		{		"BASIA01",			"BasiliskAmmo"			},
		{		"IONAMM1",			"IonAmmo"				},
		{		"D95AA0",			"ThunderAmmo"			},
		{		"GAUSICO",			"GaussRound"			},
		{		"",					""						}
	},
	// category 5 -- soul ammo
	{
		{		"AMM3A0",			"Souls"					},
		{		"",					""						}
	}
};

ammo_info_T AmmoInfo[MAX_SLOTS][MAX_AMMOTYPES_PER_SLOT] = {
	// category 1 -- clip box
	{
		{ 		200,		50 		},
		{ 		175,		43		},
		{ 		50,			12		},
		{ 		200,		50		},
		{ 		160, 		40		},
		{ 		160,		40		},
		{ 		240,		48		},
		{ 		72,			18		},
		{ 		75,			25		},
		{ 		60,			12		},
		{ 		30,			6		}
	},
	// category 2 -- shell box
	{
		{ 		50,			20		},
		{ 		90,			36		},
		{ 		60,			24		},
		{ 		35,			15		},
		{ 		40,			15		},
		{ 		60,			25		},
		{		60,			25		},
		{ 		32,			8		},
		{ 		36,			9		},
		{ 		40,			15		},
		{ 		40,			15		}
	},
	// category 3 -- rocket box
	{
		{ 		50,			5		},
		{ 		40,			10		},
		{ 		40,			4		},
		{  		50,			5		},
		{ 		25,			2		},
		{  		50,			5		},
		{ 		80,			6		},
		{ 		45,			3		},
		{ 		-1,			0		}
	},
	// category 4 -- cell box
	{
		{		300,		100		},
		{ 		270,		90		},
		{ 		240,		80		},
		{ 		250,		75		},
		{ 		375,		105		},
		{ 		175,		40		},
		{ 		250,		16		},
		{ 		180,		60		},
		{ 		200,		66		},
		{ 		75,			10		},
		{		-1,			0		}
	},
	// category 5 -- soul ammo
	{
		{ 		75,			0		},
		{ 		-1,			0		}
	}
};

enum {
	DND_MAGAZINE_SPECIALAMMOMODE_3,
	DND_MAGAZINE_SPECIALAMMOMODE_3X,
	DND_MAGAZINE_SPECIALAMMOMODE_4,
	DND_MAGAZINE_SPECIALAMMOMODE_5,
	DND_MAGAZINE_SPECIALAMMOMODE_5X,
	DND_MAGAZINE_GOREMODE,
	DND_MAGAZINE_BLADECHARGE,
	DND_MAGAZINE_SAWEDOFFCOUNTER,
	DND_MAGAZINE_AKIMBOCLIPLEFT,
	DND_MAGAZINE_AKIMBOCLIPRIGHT,
	DND_MAGAZINE_BULLETSIZE_6,
	DND_MAGAZINE_SHELLSIZE_2,
	DND_MAGAZINE_SHELLSIZE_8,
	DND_MAGAZINE_SHELLSIZE_8N,
	DND_MAGAZINE_SHELLSIZE_10,
	DND_MAGAZINE_SHELLSIZE_12,
	DND_MAGAZINE_SHELLSIZE_16,
	DND_MAGAZINE_SHELLSIZE_18,
	DND_MAGAZINE_MGCLIP,
	DND_MAGAZINE_MGCLIP2,
	DND_MAGAZINE_MGCLIP3,
	DND_MAGAZINE_MGCLIP4,
	DND_MAGAZINE_MGCLIP5,
	DND_MAGAZINE_MGCLIP6,
	DND_MAGAZINE_MGCLIP7,
	DND_MAGAZINE_FUELCLIP,
	DND_MAGAZINE_LOADEDBASILISK,
	DND_MAGAZINE_PCANCLIP,
	DND_MAGAZINE_RIOTGUNCLIP,
	DND_MAGAZINE_ACIDCLIP,
	DND_MAGAZINE_HEAVYGLCOUNTER,
	DND_MAGAZINE_INCINERATOR, // not really a mag, just put here for db save
	DND_MAGAZINE_DARKLANCE // not really a mag, just put here for db save
};

#define MAXCLIPAMMOTYPES (DND_MAGAZINE_DARKLANCE + 1)
str ClipAmmoTypes[MAXCLIPAMMOTYPES] = {
	"SpecialAmmoMode_3",
	"SpecialAmmoMode_3X",
	"SpecialAmmoMode_4",
	"SpecialAmmoMode_5",
	"SpecialAmmoMode_5X",
	"GoreMode",
	"BladeCharge",
	"SawedoffCounter",
	"AkimboClipLeft",
	"AkimboClipRight",
	"BulletSize_6",
	"ShellSize_2",
	"ShellSize_8",
	"ShellSize_8N",
	"ShellSize_10",
	"ShellSize_12",
	"ShellSize_16",
	"ShellSize_18",
	"MGClip",
	"MGClip2",
	"MGClip3",
	"MGClip4",
	"MGClip5",
	"MGClip6",
	"MGClip7",
	"FuelClip",
	"LoadedBasilisk",
	"PCanClip",
	"RiotgunClip",
	"AcidClip",
	"HeavyGLCounter",
	"IncineratorStacks",
	"LanceStacks"
};

// gets you the percentage to increase by
int GetAmmoCapIncrease() {
	int bpcount = CheckInventory("BackpackCounter");
	return (((100 + bpcount * DND_BACKPACK_RATIO)) * (100 + (GetPlayerAttributeValue(PlayerNumber(), INV_AMMOCAP_INCREASE) + GetDataFromOrbBonus(PlayerNumber(), OBI_HOLDING, -1)))) / 100;
}

bool IsBackpackLimitReached() {
	return CheckInventory("BackpackCounter") >= DND_MAX_BACKPACK;
}

void SetAllAmmoCapacities() {
	// last slot is for souls, we don't increase it here
	int factor = GetAmmoCapIncrease();
	for(int i = 0; i < MAX_SLOTS - 1; ++i)
		for(int j = 0; j < MAX_AMMOTYPES_PER_SLOT && AmmoInfo[i][j].initial_capacity != -1; ++j)
			SetAmmoCapacity(AmmoInfo_Str[i][j][AMMOINFO_NAME], (AmmoInfo[i][j].initial_capacity * factor) / 100); //Check backpack orbs.
}

void SetAllAmmoCapacitiesToDefault() {
	// last slot is for souls, we don't increase it here
	for(int i = 0; i < MAX_SLOTS - 1; ++i)
		for(int j = 0; j < MAX_AMMOTYPES_PER_SLOT && AmmoInfo[i][j].initial_capacity != -1; ++j)
			SetAmmoCapacity(AmmoInfo_Str[i][j][AMMOINFO_NAME], AmmoInfo[i][j].initial_capacity);
}

bool CheckAmmoPickup(int slot, bool simple) {
	bool res = CheckInventory(AmmoInfo_Str[slot][0][AMMOINFO_NAME]) >= GetAmmoCapacity(AmmoInfo_Str[slot][0][AMMOINFO_NAME]);
	if(!simple) {
		for(int i = 0; i < MAX_AMMOTYPES_PER_SLOT && AmmoInfo[slot][i].initial_capacity != -1; ++i)
			res = res && CheckInventory(AmmoInfo_Str[slot][i][AMMOINFO_NAME]) >= GetAmmoCapacity(AmmoInfo_Str[slot][i][AMMOINFO_NAME]);
	}
	return res;
}

int GetAmmoGainFactor() {
	return (100 + GetPlayerAttributeValue(PlayerNumber(), INV_AMMOGAIN_INCREASE) + (CheckInventory("Perk_Munitionist") * DND_MUNITION_GAIN));
}

void HandleAmmoContainerPickup(int slot, int basic_kind) {
	int amt = 0, factor = GetAmmoGainFactor();
	if (basic_kind > 0) {
			amt = AmmoInfo[slot][0].container_value; //large pack
			if (basic_kind <= 2)
				amt /= 5; //small pack
			if (basic_kind == 1)
				amt /= 2; //dropped pack (clip only)
		
			amt = amt * factor / 100;
			if(!amt)
				amt = 1;

			GiveInventory(AmmoInfo_Str[slot][0][AMMOINFO_NAME], amt);
	} else {
		for(int i = 0; i < MAX_AMMOTYPES_PER_SLOT && AmmoInfo[slot][i].initial_capacity != -1; ++i) {
			amt = AmmoInfo[slot][i].container_value;
			amt = amt * factor / 100;
			if(!amt)
				amt = 1;

			GiveInventory(AmmoInfo_Str[slot][i][AMMOINFO_NAME], amt);
		}
	}
}

void GiveAmmo(int amt, int slot, int t) {
	if(slot != DND_AMMOSLOT_SOULS)
		amt = amt * GetAmmoGainFactor() / 100;
	else
		amt = amt + (amt * CheckInventory("IATTR_SoulAmmoIncrease") / 100);
	GiveInventory(AmmoInfo_Str[slot][t][AMMOINFO_NAME], amt);
}

#endif
